aspnetcore/blazor/security/webassembly/additional-scenarios.md

```diff
@@ -1008,3 +1008,65 @@ Alternatively, the setting can be made in the app settings (`appsettings.json`)
 If tacking on a segment to the authority isn't appropriate for the app's OIDC provider, such as with non-AAD providers, set the <xref:Microsoft.AspNetCore.Builder.OpenIdConnectOptions.Authority> property directly. Either set the property in <xref:Microsoft.AspNetCore.Builder.JwtBearerOptions> or in the app settings file (`appsettings.json`) with the `Authority` key.
 
 The list of claims in the ID token changes for v2.0 endpoints. For more information, see [Why update to Microsoft identity platform (v2.0)?](/azure/active-directory/azuread-dev/azure-ad-endpoint-comparison).
+
+## Configure and use gRPC in components
+
+To configure a Blazor WebAssembly app to use the [ASP.NET Core gRPC framework](xref:grpc/index), register gRPC services for the app's message handler. The following example configures the app's authorization message handler to use the [`GreeterClient` service from the gRPC tutorial](xref:tutorials/grpc/grpc-start#create-a-grpc-service) (`Program.Main`):
+
+```csharp
+using System.Net.Http;
+using Microsoft.AspNetCore.Components.WebAssembly.Authentication;
+using Grpc.Net.Client;
+using Grpc.Net.Client.Web;
+using {APP ASSEMBLY}.Shared;
+
+...
+
+builder.Services.AddScoped(sp =>
```

Comment:
Any reason for AddScoped over AddSingleton?

You generally want to reuse a `GrpcChannel` instance for perf benefits.

Reply:
It came from the sample app at ...

https://github.com/javiercn/BlazorGrpcAuth/blob/master/BlazorGrpcAuth/Client/Program.cs#L30

Reply:
@javiercn Is there a reason to use scope lifetime here? Does `BaseAddressAuthorizationMessageHandler` have a scoped lifetime?

Reply:
Transient ...

https://github.com/dotnet/aspnetcore/blob/648c15dbe90fb8c113f7c6b4adeb40d9e10494f6/src/Components/WebAssembly/WebAssembly.Authentication/src/WebAssemblyAuthenticationServiceCollectionExtensions.cs#L41

Reply:
@JamesNK So that it can work in Server-Side contexts (prerendering) without ugly surprises.

Reply:
If we change it to Singleton it throws:
```
Microsoft.AspNetCore.Components.WebAssembly.Rendering.WebAssemblyRenderer[100]
      Unhandled exception rendering component: Cannot resolve 'Microsoft.AspNetCore.Components.WebAssembly.Authentication.BaseAddressAuthorizationMessageHandler' from root provider because it requires scoped service 'Microsoft.AspNetCore.Components.WebAssembly.Authentication.IAccessTokenProvider'.
System.InvalidOperationException: Cannot resolve 'Microsoft.AspNetCore.Components.WebAssembly.Authentication.BaseAddressAuthorizationMessageHandler' from root provider because it requires scoped service 'Microsoft.AspNetCore.Components.WebAssembly.Authentication.IAccessTokenProvider'.
```

Reply:
Yeah, forgot about it. The ATP is scoped.

Reply:
Are we good ... can I merge this now?

Reply:
The referenced sample code does not work, BTW.
https://github.com/javiercn/BlazorGrpcAuth/tree/master/BlazorGrpcAuth

It throws an exception.

Reply:
We should add exception handling to the code, IMO. For the cases when there is no token available.

Reply:
Usually, we are supposed to catch an AccessTokenNotAvailableException in Blazor and then e.g. call .Redirect to get redirected to the IdP..
However, when using a gRPC channel in Blazor we only get a RpcException, with no inner exception to act upon.

Maybe @JamesNK could add some thoughts here?

Reply:
Should error handling be added? Related issue - https://github.com/grpc/grpc-dotnet/issues/965